R/read_csv.R

```diff
@@ -0,0 +1,239 @@
+#' it returns, the sampling informations match.
+#'
+#' @param a the first sampling info to check
+#' @param b the second sampling info to check
+#'
+check_sampling_csv_info_matches <- function(a, b) {
+  if(a$stan_version_major != b$stan_version_major ||
+     a$stan_version_minor != b$stan_version_minor ||
+     a$stan_version_patch != b$stan_version_patch) {
+    stop("Supplied CSV files were not generated with the same version of Cmdstan!")
+  }
+  if(a$model != b$model) {
+    stop("Supplied CSV files were not generated wtih the same model!")
+  }
+  if(!all(a$params == b$params)) {
+    stop("Supplied CSV files have samples for different parameters!")
+  }
+  if(a$data_file != b$data_file) {
+    stop("Supplied CSV files have samples from chains run with non-matching data!")
+  }
+  for(name in names(a)) {
+    if(startsWith(name, "sample_")) {
+      if (is.null(b[[name]]) ||
+          a[[name]] != b[[name]]) {
+        stop("Supplied CSV files do not match in all sampling settings!")
+      }
+    }
+  }
+}
+
+#' Reads the sampling arguments and the diagonal of the
+#' inverse mass matrix from the comments in a CSV file.
+#'
+#' @param csv_file A CSV file containing results from sampling
+#' @return A list containing all sampling parameters and the
+#' diagonal of the inverse mass matrix
+#'
+read_sample_info_csv <- function(csv_file) {
+  param_names_read <- FALSE
+  sampling_params_read <- FALSE
+  diagonal_matrix_next <- FALSE
+  arg_prefix <- ""
+  csv_file_info = list()
+  con  <- file(csv_file, open = "r")
+  while (length(line <- readLines(con, n = 1, warn = FALSE)) > 0) {
+    if (!startsWith(line, "#")) {
+      if(!param_names_read) {
+        param_names_read <- TRUE
+        all_names <- strsplit(line, ",")[[1]]
+        csv_file_info[["sampler_params"]] <- c()
+        csv_file_info[["model_params"]] <- c()
+        for(x in all_names) {
+          if(endsWith(x, "__") && x != "lp__"){
+            csv_file_info[["sampler_params"]] <- c(csv_file_info[["sampler_params"]], x)
+          } else {
+            csv_file_info[["model_params"]] <- c(csv_file_info[["model_params"]], x)
+          }
+        }
+        next;
+      } else {
+        break;
+      }
+    }
+    # no more sampling settings follow the list of parameters
+    if(!param_names_read) {
+      tmp <- gsub("#", "", line, fixed = TRUE)
+      tmp <- gsub("(Default)", "", tmp, fixed = TRUE)
+      key_val <- grep("=", tmp, fixed = TRUE, value = TRUE)
+      if(length(key_val) == 0) {
+        arg_name <- trimws(tmp)
+        if(arg_prefix != "" && !sampling_params_read) {
+          arg_prefix <- paste(arg_prefix, arg_name, sep = "_")
+        } else {
+          arg_prefix <- trimws(tmp)
+        }
+      } else {
+        key_val <- strsplit(key_val, split = "=", fixed = TRUE)
+        key_val <- rapply(key_val, trimws)
+        if(length(key_val) == 2) {
+          numeric_val <- suppressWarnings(as.numeric(key_val[2]))
```

Comment:
Is there a better way of checking whether a string can be converted to a number?

For now I do this and then is.na.

Reply:
That's fine. Not sure if there's a better way.